Add tool name filtering to `getAgentTools()`

peteski22 opened the pull request:
Adds `tools` parameter to `getAgentTools()` for filtering by tool name.

Supports two formats:
- Raw: `tools: ['add']` - matches tool across all servers
- Prefixed: `tools: ['time__get_current_time']` - matches specific server+tool

Can be combined with server filtering for progressive disclosure (operators filter servers, agents filter tools).

Introduces `TOOL_SEPARATOR` constant and validation for prefixed format.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added client-side tool-name filtering when retrieving agent tools; accepts raw names or server-prefixed identifiers and can be combined with server and format options.

* **API**
  * Public API updated to accept an optional tools?: string[] option alongside servers? and format?.

* **Documentation**
  * README updated with cross-server, server-qualified and mixed-filter examples and notes on output formats (default: array).

* **Tests**
  * New unit tests covering varied filtering scenarios and all output formats.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

peteski22 committed:
Add tool name filtering to getAgentTools()

Enables filtering tools by name in addition to server-level filtering.
Supports both raw tool names (e.g., 'add') and server-prefixed names
(e.g., 'time__get_current_time').

- Add tools parameter to AgentToolsOptions interface
- Implement filtering logic with validation for prefixed format
- Add TOOL_SEPARATOR constant to eliminate magic strings
- Add comprehensive unit tests (9 test cases)
- Update README with examples and progressive disclosure explanation

khaledosman commented on src/client.ts:
```diff
@@ -789,36 +796,72 @@ export class McpdClient {
   async getAgentTools(options?: {
     format?: "array";
     servers?: string[];
+    tools?: string[];
   }): Promise<AgentFunction[]>;
   async getAgentTools(options: {
     format: "object";
     servers?: string[];
+    tools?: string[];
   }): Promise<Record<string, AgentFunction>>;
   async getAgentTools(options: {
     format: "map";
     servers?: string[];
+    tools?: string[];
```
nit: instead of making `tools` optional with `?` which makes it type `string[] | undefined` why not just give it a default value of `[]` instead to maintain a single type without worrying about `undefined` checks?

khaledosman commented on src/client.ts:
```diff
@@ -789,36 +796,72 @@ export class McpdClient {
+        const leftPart = filterItem.slice(0, separatorIndex).trim();
+        const rightPart = filterItem
+          .slice(separatorIndex + TOOL_SEPARATOR.length)
+          .trim();
+
+        if (leftPart && rightPart && filterItem === tool.name) {
+          return true;
+        }
+      }
+
+      return filterItem === tool._toolName;
+    });
```
would something like this work?


```suggestion
  return tools.some((toolFilter) => {
    if (toolFilter.includes(TOOL_SEPARATOR)) {
      return toolFilter.trim() === tool.name;
    } else {
      return toolFilter === tool._toolName;
    }
  });
```

peteski22 replied:
Will do something like this (thanks!) the current logic I added was supposed to be to show that we validated the left/right bits, but TBH it's not super required.

peteski22 committed:
Simplify tool filter validation logic and add ambiguous filter test

- Remove unnecessary validation in #matchesToolFilter
- Malformed filter strings naturally fail to match
- Add test case for ambiguous filter names
